snapcraft/tests/test_parser.py

```diff
@@ -556,3 +561,41 @@ def test_missing_fields(self, mock_get, mock_get_origin_data):
         }
         main(['--debug', '--index', TEST_OUTPUT_PATH])
         self.assertEqual(0, _get_part_list_count())
+
+    def test__get_origin_data_both(self):
+        tempdir = tempfile.mkdtemp()
+        with open(os.path.join(tempdir, '.snapcraft.yaml'), 'w') as fp:
+            fp.write("")
+        with open(os.path.join(tempdir, 'snapcraft.yaml'), 'w') as fp:
+            fp.write("")
+
+        self.assertRaises(BadSnapcraftYAMLError, _get_origin_data, tempdir)
+        shutil.rmtree(tempdir)
+
+    def test__get_origin_data_hidden_only(self):
+        tempdir = tempfile.mkdtemp()
+        with open(os.path.join(tempdir, '.snapcraft.yaml'), 'w') as fp:
+            fp.write("")
+
+        _get_origin_data(tempdir)
+        shutil.rmtree(tempdir)
+
+    def test__get_origin_data_normal_only(self):
+        tempdir = tempfile.mkdtemp()
+        with open(os.path.join(tempdir, 'snapcraft.yaml'), 'w') as fp:
+            fp.write("")
+
+        _get_origin_data(tempdir)
+        shutil.rmtree(tempdir)
+
+    def test__encode_origin_git(self):
+        origin = 'git@github.com:testuser/testproject.git'
+        origin_dir = _encode_origin(origin)
+
+        self.assertEqual('gitgithub.comtestusertestproject.git', origin_dir)
+
+    def test__encode_origin_lp(self):
+        origin = 'lp:~testuser/testproject/testbranch'
+        origin_dir = _encode_origin(origin)
+
+        self.assertEqual('lptestusertestprojecttestbranch', origin_dir)
```

Comment:
That's the thing. We should be able to easily inject a fake server or client with a canned or seeded response that will reply quickly. That will take a lot of time once, but then the following tests will be easier. We could also move some of these internal methods into a module, and make them the public api of that module, which parser consumes. That could be cheating if we are only doing that to make them look less private, but a combination of both could be nice, and we can leave a few whitebox tests for hard details. Again, I'm not blocking this branch because of this, just starting the discussion to find a way to make future tests a little less whitebox.